include: zephyr: drivers: modify SPI_CONFIG_DT to add CPOL, CPHA, HOLD_CS

HoZHel opened the pull request:
Expose SPI device hardware characteristics: clock polarity, phase, and ability to control the chip select in application via DeviceTree files.
This allows users to define these SPI hardware characteristics in dts, overlay, etc. files without the need to modify the C code.

HoZHel committed:
drivers: spi: update drivers that were using spi cpol and cpha

Modified files (yaml, dts, overlay, and c) which were using spi-cpol
and spi-cpha to be compatible with the new structure.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

simonguinot commented on include/zephyr/drivers/spi.h:
```diff
@@ -102,8 +86,6 @@ extern "C" {
  * @{
  */
-/* Requests - if possible - to keep CS asserted after the transaction */
-#define SPI_HOLD_ON_CS		BIT(12)
```
Maybe you should move the removals above into the next patch:  `include: zephyr: dt-bindings: spi: move CPOL, CPHA, HOLD_CS declarations`. Else all these definitions are missing between the two patches and this could break git-bisect.

HoZHel replied:
Thanks for your comment. I merged the 2 commits.

simonguinot commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -34,8 +34,6 @@ LOG_MODULE_REGISTER(ws2812_spi);
  */
 #define SPI_OPER(idx) (SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | \
-		  COND_CODE_1(DT_INST_PROP(idx, spi_cpol), (SPI_MODE_CPOL), (0)) | \
-		  COND_CODE_1(DT_INST_PROP(idx, spi_cpha), (SPI_MODE_CPHA), (0)) | \
```
The changes are looking OK regarding the `ws2812_spi` driver. I'll try to test.

mbolivar-ampere reviewed (changes requested):
Linux is using flags for `spi-cpol` and `spi-cpha`, not enums, and we try to preserve compatibility with linux for standard properties. If there is no strong justification for doing this differently, please match Linux. If there is a strong justification, please state it.

Regarding spi-hold-cs, is this really necessary at the spi-device.yaml level, and why must it be an enum when it is essentially a flag also?

HoZHel commented:
> Linux is using flags for `spi-cpol` and `spi-cpha`, not enums, and we try to preserve compatibility with linux for standard properties. If there is no strong justification for doing this differently, please match Linux. If there is a strong justification, please state it.
>

@mbolivar-ampere, I don't have a strong reason, just I did that to follow the same style in spi-device.yaml, and I was not aware of the Linux style, so I will change them to boolean.
 
> Regarding spi-hold-cs, is this really necessary at the spi-device.yaml level, and why must it be an enum when it is essentially a flag also?

Some products in the market, such as ST BlueNRG series require hold-cs feature. According to Zephyr policy, hardware configurations should be in devicetree files, and in order to use it in dts, overlay, etc. files, it is necessary to be defined as a property in a yaml file.
In my opinion, since spi-device.yaml has fundamental SPI properties, it's the best file to include these flags in comparison to other files, like zephyr,bt-hci-spi.yaml.

In addition, the purpose of using enum was the same as above, so I will also change this to boolean.

mbolivar-ampere commented:
> In addition, the purpose of using enum was the same as above, so I will also change this to boolean.

Ack, thanks!

HoZHel committed:
include: zephyr: drivers: modify SPI_CONFIG_DT to add CPOL, CPHA, HOLD_CS

Modify SPI_CONFIG_DT macro to config SPI clock polarity (CPOL),
clock phase (CPHA), and hold-cs.

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>

gmarull commented on dts/bindings/spi/spi-device.yaml:
```diff
@@ -40,3 +40,21 @@ properties:
+  spi-hold-cs:
+    type: boolean
+    description: |
+      In some cases, such as BlueNRG-MS, it is necessary for the master to manage
```
please avoid specific information in generic bindings docs, ie, what is "BlueNRG-MS"? Just explain what `spi-hold-cs` implies.

HoZHel replied:
I wanted to provide a real example; however, if you prefer not to have it, I can remove it.

gmarull replied:
I understand, but for readers without knowledge of what BlueNRG-MS is, the text doesn't add much value. I'd document in the BlueNRG-MS bindings about the need for this option, though.

HoZHel replied:
Ok. I removed it.

HoZHel committed:
dts: bindings: spi: add support to set CPOL, CPHA, and HOLD_CS in dts file

Add support to set SPI clock polarity (CPOL), clock phase (CPHA), and
hold-on-cs in a dts file to get rid of using related macros in spi.c driver
since each board may work on a different SPI mode rather than the default
one (based on CPOL and CPHA).

Signed-off-by: Ali Hozhabri <ali.hozhabri@st.com>
